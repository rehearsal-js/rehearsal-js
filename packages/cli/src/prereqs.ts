// ! these should be the bare minimum versions and config required for rehearsal to work with type inference
// ! deps are >= the version specified

import type { PreReqs, ProjectType, PreReqTSConfig } from './types.js';

// node package.json config must be gte 14
const nodeBasePrereq = '14';

// eslint
const eslintBasePrereq = '@typescript-eslint/parser';

// tsconfig
const tsconfigBasePrereq: PreReqTSConfig = {
  compilerOptions: {
    strict: true,
    skipLibCheck: true,
  },
};

// tsconfig ember app/addon
const tsconfigEmberPrereq: PreReqTSConfig = {
  ...tsconfigBasePrereq,
  glint: {
    environment: 'ember-loose',
  },
};

// tsconfig glimmer
const tsconfigGlimmerPrereq: PreReqTSConfig = {
  ...tsconfigBasePrereq,
  glint: {
    environment: 'glimmerx',
  },
};

// pre-req dependencies for rehearsal. doesn't matter if they are deps or devDeps
const depsBasePreReq = {
  typescript: '4.9.0',
  prettier: '2.0.0',
  eslint: '8.0.0',
  'eslint-import-resolver-typescript': '2.5.0',
  'eslint-plugin-import': '2.0.0',
  'eslint-plugin-node': '11.0.0',
  'eslint-plugin-unicorn': '6.0.0',
  '@typescript-eslint/eslint-plugin': '5.0.0',
  '@typescript-eslint/parser': '5.0.0',
};

// pre-req dependencies for ember app/addon
const depsEmberPreReq = {
  ...depsBasePreReq,
  '@glint/core': '1.0.0',
  '@glint/environment-ember-loose': '1.0.0',
  '@glint/environment-ember-template-imports': '1.0.0',
  '@glint/template': '1.0.0',
  'eslint-plugin-ember': '11.0.0',
  'prettier-plugin-ember-template-tag': '0.3.0',
};

// pre-req dependencies for glimmer
const depsGlimmerPreReq = {
  ...depsBasePreReq,
  '@glint/core': '1.0.0',
  '@glint/environment-glimmerx': '1.0.0',
  '@glint/template': '1.0.0',
  'eslint-plugin-ember': '11.0.0',
  '@glimmerx/prettier-plugin-component-templates': '0.6.0',
};

const preReqs: Record<string, PreReqs> = {
  'base-ts': {
    node: nodeBasePrereq,
    eslint: eslintBasePrereq,
    tsconfig: {
      ...tsconfigBasePrereq,
    },
    deps: {
      ...depsBasePreReq,
    },
  },
  ember: {
    node: nodeBasePrereq,
    eslint: eslintBasePrereq,
    tsconfig: {
      ...tsconfigEmberPrereq,
    },
    deps: {
      ...depsEmberPreReq,
    },
  },
  glimmer: {
    node: nodeBasePrereq,
    eslint: eslintBasePrereq,
    tsconfig: {
      ...tsconfigGlimmerPrereq,
    },
    deps: {
      ...depsGlimmerPreReq,
    },
  },
};

export function getPreReqs(type: ProjectType): PreReqs {
  return preReqs[type];
}
