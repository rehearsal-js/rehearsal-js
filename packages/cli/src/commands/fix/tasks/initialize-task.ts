import { resolve } from 'node:path';
import { ListrDefaultRenderer, ListrTask } from 'listr2';
import debug from 'debug';
import { isGlintProject, isAddon, isApp } from '@rehearsal/service';
import { readJsonSync } from 'fs-extra/esm';
import { PackageJson } from 'type-fest';
import { getPreReqs } from '../../../prereqs.js';
import { validateSourcePath } from '../../../utils/paths.js';

import {
  determineProjectName,
  isDepsPreReq,
  isESLintPreReq,
  isExistsESLintConfig,
  isExistsPackageJSON,
  isExistsTSConfig,
  isNodePreReq,
  isTSConfigPreReq,
  isValidGitIgnore,
} from '../../../utils/prereq-checks.js';
import type { FixCommandOptions, CommandContext, ProjectType, SkipChecks } from '../../../types.js';

const DEBUG_CALLBACK = debug('rehearsal:cli:fix:init-task');

// everything is relative to the project root. options.basePath cannot be configured by the user
export function initTask(
  srcPath: string,
  options: FixCommandOptions,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _ctx?: CommandContext
): ListrTask<CommandContext, ListrDefaultRenderer> {
  return {
    title: `Initialize`,
    task: async (ctx: CommandContext): Promise<void> => {
      const { rootPath, skipDepsCheck, skipEsLintCheck } = options;

      const skipChecks = {
        skipDepsCheck,
        skipEsLintCheck,
      };

      let projectType: ProjectType = 'base-ts';
      const packageJSON = readJsonSync(resolve(rootPath, 'package.json')) as PackageJson;
      ctx.projectName = determineProjectName(rootPath) || '';

      // if ember app or addon
      if (isApp(packageJSON) || isAddon(packageJSON)) {
        projectType = 'ember';
      } else if (await isGlintProject(rootPath)) {
        projectType = 'glimmer';
      }

      preFlightCheck(rootPath, projectType, skipChecks);

      // in this mode we skip the `graphOrderTask`
      if (process.env['GRAPH_MODES'] === 'off') {
        // grab all the ts files since we don't have graph output
        [ctx.orderedFiles] = validateSourcePath(rootPath, srcPath, 'ts');
      }

      DEBUG_CALLBACK('ctx %O:', ctx);
    },
    options: {
      bottomBar: false,
      persistentOutput: false,
    },
  };
}

// each of these sub-tasks will throw on failure
export function preFlightCheck(
  basePath: string,
  projectType: ProjectType,
  skipChecks: SkipChecks,
  isSkipped = false
): void {
  // FOR LOCAL DEVELOPMENT TESTING ONLY! SKIP ALL PRE-REQ CHECKS
  if (isSkipped) {
    return;
  }

  const { deps, eslint, tsconfig, node } = getPreReqs(projectType);

  // is exists checks these will throw faster than the prereq checks
  isExistsPackageJSON(basePath);
  isExistsTSConfig(basePath);
  isExistsESLintConfig(basePath);
  isValidGitIgnore(basePath);
  // prereq checks for both the version and the package
  // if `--skipDepsCheck` flag is true skip `isDepsPreReq`
  if (!skipChecks.skipDepsCheck) {
    isDepsPreReq(basePath, deps);
  }
  // if `--skipEsLintCheck` flag is true skip `isEslintPreReq`
  if (!skipChecks.skipEsLintCheck) {
    isESLintPreReq(basePath, eslint);
  }
  isTSConfigPreReq(basePath, tsconfig);
  isNodePreReq(node);
}
