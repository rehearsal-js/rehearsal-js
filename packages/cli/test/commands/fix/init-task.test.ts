import { resolve } from 'node:path';
import { Project } from 'fixturify-project';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { createLogger, format, transports } from 'winston';

import {
  prepareProject,
  listrTaskRunner,
  cleanOutput,
  createOutputStream,
} from '../../test-helpers/index.js';
import { preFlightCheck } from '../../../src/commands/fix/tasks/initialize-task.js';
import { getPreReqs } from '../../../src/prereqs.js';
import { initTask } from '../../../src/commands/fix/tasks/index.js';

import type {
  ProjectType,
  CommandContext,
  FixCommandOptions,
  SkipChecks,
} from '../../../src/types.js';
import type { Readable } from 'node:stream';
import type { DirJSON } from 'fixturify';

const logger = createLogger({
  transports: [new transports.Console({ format: format.cli() })],
});

function projectAddDevDeps(project: Project, type: ProjectType): void {
  const prereqs = getPreReqs(type);
  const deps = prereqs.deps;
  for (const [dep, version] of Object.entries(deps)) {
    project.addDevDependency(dep, `^${version}`);
  }

  project.linkDependency('typescript', { baseDir: __dirname });
}

function projectInit(project: Project, type: ProjectType): void {
  projectAddDevDeps(project, type);

  const prereqs = getPreReqs(type);

  project.files['tsconfig.json'] = JSON.stringify({
    ...prereqs.tsconfig,
  });
  project.files['.eslintrc.json'] = JSON.stringify({
    parser: prereqs.eslint,
  });
}

function submoduleInit(project: Project, type: ProjectType, submoduleName: string): void {
  projectAddDevDeps(project, type);

  const prereqs = getPreReqs(type);

  (project.files[submoduleName] as DirJSON)['tsconfig.json'] = JSON.stringify({
    ...prereqs.tsconfig,
  });
}

describe('Fix: Init-Task', () => {
  let project: Project;
  let outputStream: Readable;
  let output = '';

  vi.spyOn(console, 'info').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
  });
  vi.spyOn(console, 'log').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
  });
  vi.spyOn(console, 'error').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
  });
  vi.spyOn(logger, 'warn').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
    return logger;
  });
  vi.spyOn(logger, 'error').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
    return logger;
  });

  beforeEach(async () => {
    output = '';
    outputStream = createOutputStream();
    project = prepareProject('base_js_app');
    await project.write();
  });

  afterEach(() => {
    output = '';
    vi.clearAllMocks();
    project.dispose();
  });

  test(`preFlightCheck "base-ts" works`, async () => {
    const projectType: ProjectType = 'base-ts';

    projectInit(project, projectType);
    // support "latest" as a version
    project.addDevDependency('typescript', 'latest');

    await project.write();

    try {
      preFlightCheck(project.baseDir, project.baseDir, projectType);
    } catch (error) {
      expect(error).toBeUndefined();
    }

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });

  test(`preFlightCheck "ember" works`, async () => {
    const projectType: ProjectType = 'ember';

    projectInit(project, projectType);

    await project.write();

    try {
      preFlightCheck(project.baseDir, project.baseDir, projectType);
    } catch (error) {
      expect(error).toBeUndefined();
    }

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });

  test(`preFlightCheck "ember" submodule of a "base-ts" module`, async () => {
    projectInit(project, 'base-ts');

    submoduleInit(project, 'ember', 'module-c');

    await project.write();

    const srcPath = resolve(project.baseDir, 'module-c');

    try {
      preFlightCheck(srcPath, project.baseDir, 'ember');
    } catch (error) {
      expect(error).toBeUndefined();
    }

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });

  test(`preFlightCheck "glimmer" works`, async () => {
    const projectType: ProjectType = 'glimmer';

    projectInit(project, projectType);

    await project.write();

    try {
      preFlightCheck(project.baseDir, project.baseDir, projectType);
    } catch (error) {
      expect(error).toBeUndefined();
    }

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });

  test(`preFlightCheck "glimmer" submodule of a "ember" module`, async () => {
    projectInit(project, 'ember');

    submoduleInit(project, 'glimmer', 'module-c');

    await project.write();

    const srcPath = resolve(project.baseDir, 'module-c');

    try {
      preFlightCheck(srcPath, project.baseDir, 'glimmer');
    } catch (error) {
      expect(error).toBeUndefined();
    }

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });

  test('preFlightCheck "base-ts" - deps/devDeps compare', async () => {
    const projectType: ProjectType = 'base-ts';

    projectInit(project, projectType);

    // remove from devDeps and add them to deps
    project.removeDevDependency('typescript');
    project.addDependency('typescript', '^5.0.0');

    await project.write();

    try {
      preFlightCheck(project.baseDir, project.baseDir, projectType);
    } catch (error) {
      expect(error).toBeUndefined();
    }

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });

  test('preFlightCheck "base-ts" - deps/devDeps compare in submodule', async () => {
    const projectType: ProjectType = 'base-ts';

    projectInit(project, projectType);

    // remove from devDeps and add them to deps
    project.removeDevDependency('typescript');
    project.addDependency('typescript', '^5.0.0');

    await project.write();

    const srcPath = resolve(project.baseDir, 'module-c');

    try {
      preFlightCheck(srcPath, project.baseDir, projectType);
    } catch (error) {
      expect(error).toBeUndefined();
    }

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });

  test('preFlightCheck "base-ts" - expect failure tsconfig strict as false', async () => {
    const projectType: ProjectType = 'base-ts';
    const prereqs = getPreReqs(projectType);
    project.files['tsconfig.json'] = JSON.stringify({
      compilerOptions: {
        strict: false,
        skipLibCheck: true,
      },
    });
    project.files['.eslintrc.json'] = JSON.stringify({
      parser: prereqs.eslint,
    });

    // adds all the deps from the prereqs for the given `ProjectType`
    projectAddDevDeps(project, projectType);

    await project.write();

    try {
      preFlightCheck(project.baseDir, project.baseDir, projectType);
    } catch (error) {
      expect(error).toBeDefined();
      if (error instanceof Error) {
        expect(cleanOutput(error.message, project.baseDir)).toMatchSnapshot();
      }
    }

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });

  test('preFlightCheck "ember" - expect failure tsconfig glint missing', async () => {
    const projectType: ProjectType = 'ember';

    project.files['tsconfig.json'] = JSON.stringify({
      compilerOptions: {
        strict: true,
        skipLibCheck: true,
      },
    });
    project.files['.eslintrc.json'] = JSON.stringify({
      parser: '@typescript-eslint/parser',
    });

    // adds all the deps from the prereqs for the given `ProjectType`
    projectAddDevDeps(project, projectType);

    await project.write();

    try {
      preFlightCheck(project.baseDir, project.baseDir, projectType);
    } catch (error) {
      expect(error).toBeDefined();
      if (error instanceof Error) {
        expect(cleanOutput(error.message, project.baseDir)).toMatchSnapshot();
      }
    }

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });

  test('preFlightCheck "base-ts" - expect failure eslint parser invalid', async () => {
    const projectType: ProjectType = 'base-ts';

    project.files['tsconfig.json'] = JSON.stringify({
      compilerOptions: {
        strict: true,
        skipLibCheck: true,
      },
    });
    project.files['.eslintrc.json'] = JSON.stringify({
      parser: '@fake/parser',
    });

    // adds all the deps from the prereqs for the given `ProjectType`
    projectAddDevDeps(project, projectType);

    await project.write();

    try {
      preFlightCheck(project.baseDir, project.baseDir, projectType);
    } catch (error) {
      expect(error).toBeDefined();
      if (error instanceof Error) {
        expect(cleanOutput(error.message, project.baseDir)).toMatchSnapshot();
      }
    }

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });

  test('preFlightCheck "base-ts" - expect failure deps missing', async () => {
    const projectType: ProjectType = 'base-ts';

    projectInit(project, projectType);

    project.removeDevDependency('typescript');
    project.removeDevDependency('prettier');
    project.removeDevDependency('eslint');

    await project.write();

    try {
      preFlightCheck(project.baseDir, project.baseDir, projectType);
    } catch (error) {
      expect(error).toBeDefined();
      if (error instanceof Error) {
        expect(cleanOutput(error.message, project.baseDir)).toMatchSnapshot();
      }
    }

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });

  test('preFlightCheck "base-ts" - no failure when deps missing and `skipChecks deps` option is set', async () => {
    const projectType: ProjectType = 'base-ts';
    const skipChecks: SkipChecks = ['deps'];

    projectInit(project, projectType);

    project.removeDevDependency('typescript');
    project.removeDevDependency('prettier');
    project.removeDevDependency('eslint');

    await project.write();

    try {
      preFlightCheck(project.baseDir, project.baseDir, projectType, skipChecks);
    } finally {
      expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
    }
  });

  test('preFlightCheck "base-ts" - no failure when deps missing and `skipChecks eslint` option is set', async () => {
    const projectType: ProjectType = 'base-ts';
    const skipChecks: SkipChecks = ['eslint'];

    projectInit(project, projectType);
    // removes the `@typescript-eslint/parser` from projects `eslintrc.json`
    project.files['eslintrc.json'] = {};

    await project.write();

    try {
      preFlightCheck(project.baseDir, project.baseDir, projectType, skipChecks);
    } finally {
      expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
    }
  });

  test(`preFlightCheck "base-ts" submodule of a "glimmer" module`, async () => {
    projectInit(project, 'glimmer');

    submoduleInit(project, 'base-ts', 'module-c');

    await project.write();

    const srcPath = resolve(project.baseDir, 'module-c');

    try {
      preFlightCheck(srcPath, project.baseDir, 'base-ts');
    } catch (error) {
      expect(error).toBeUndefined();
    }

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });

  test(`validate initTask "base-ts" works with src arg`, async () => {
    const projectType: ProjectType = 'base-ts';

    project = Project.fromDir(resolve(__dirname, '../../fixtures/base_ts_app'), {
      linkDeps: true,
      linkDevDeps: true,
    });

    projectInit(project, projectType);
    await project.write();

    const src = resolve(project.baseDir, 'src');
    const options: FixCommandOptions = {
      rootPath: project.baseDir,
      format: ['sarif'],
      ignore: [],
      mode: 'single-pass',
    };
    process.env['GRAPH_MODES'] = 'off';
    const tasks = [initTask(src, options)];
    const ctx = await listrTaskRunner<CommandContext>(tasks);
    const sanitizedAbsPaths = ctx.orderedFiles?.map((path) => {
      return cleanOutput(path, project.baseDir);
    });

    expect(sanitizedAbsPaths).toMatchSnapshot();
    expect(ctx.orderedFiles.map((file) => file.replace(project.baseDir, ''))).toMatchSnapshot();
    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
    process.env['GRAPH_MODES'] = undefined;
  });
});
