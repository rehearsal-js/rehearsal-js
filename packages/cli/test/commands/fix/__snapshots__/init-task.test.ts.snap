// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Fix: Init-Task > preFlightCheck base - expect failure deps missing 1`] = `
[Error: Please install the following missing devDependencies and try again: 
"typescript >= 5.0.0", 
"prettier >= 2.0.0", 
"eslint >= 8.0.0"]
`;

exports[`Fix: Init-Task > preFlightCheck base - expect failure eslint parser invalid 1`] = `[Error: Please update "parser: @fake/parser" to "parser: @typescript-eslint/parser" in the ESLint config and try again.]`;

exports[`Fix: Init-Task > preFlightCheck base - expect failure tsconfig strict as false 1`] = `
[Error: Please update the tsconfig.json to include:
{
  "compilerOptions": {
    "strict": true,
    "skipLibCheck": true
  }
}]
`;

exports[`Fix: Init-Task > preFlightCheck ember - expect failure tsconfig glint missing 1`] = `
[Error: The tsconfig.json is missing the following keys:
"glint".
Please update the tsconfig.json to include:
{
  "compilerOptions": {
    "strict": true,
    "skipLibCheck": true
  },
  "glint": {
    "environment": [
      "ember-loose",
      "ember-template-imports"
    ],
    "checkStandaloneTemplates": true
  }
}]
`;

exports[`Fix: Init-Task > validate initTask base works with --source 1`] = `
[
  "<tmp-path>/src/index.ts",
  "<tmp-path>/src/get-live-neighbor-count.ts",
  "<tmp-path>/src/gen-random-grid.ts",
  "<tmp-path>/src/apply-rules.ts",
]
`;

exports[`Fix: Init-Task > validate initTask base works with --source 2`] = `
[
  "src/index.ts",
  "src/get-live-neighbor-count.ts",
  "src/gen-random-grid.ts",
  "src/apply-rules.ts",
]
`;
